### PR TITLE
[WIP] Arm PerfDoc Checks 18 through 20 - Compute Shader Analysis

### PR DESCRIPTION
**Pull request by samwalls-arm**
~~This pull request has another round of changes ported from PerfDoc. Here, we add SPIRV-Cross as a dependency to the `known_good.json` file, and integrate its usage into checks for a few different performance warnings regarding compute shaders.~~

~~I've added SPIRV-Cross support in the CMakeLists.txt file, with observed Windows and Unix support. However, I'm unsure if this will work on android. Is there anything else we would need to do to make that work?~~

EDIT: I've removed SPIRV-Cross, and implemented the checks another way - see the conversation below.

### CODE REVIEW

**Comment by samwalls-arm** [comment]
Hey @mark-lunarg So, I think I got windows + unix builds working (required a small edit to enable `-fPIC`). Looks like the travis CI build for Android can't find spirv-cross includes. Where would this be configured?

File: layers/best_practices_utils.cpp [hunk 1]
```diff
@@ -855,6 +859,105 @@ bool BestPractices::PreCallValidateCreateComputePipelines(VkDevice device, VkPip
...
+}
+
+bool BestPractices::ValidateCreateComputePipelineArm(const VkComputePipelineCreateInfo& createInfo) const {
+    bool skip = false;
+    auto* module = GetShaderModuleState(createInfo.stage.module);
+    try {
+        auto compiler = std::unique_ptr<spirv_cross::Compiler>(new spirv_cross::Compiler(module->words));
+        compiler->set_entry_point(createInfo.stage.pName, spv::ExecutionModelGLCompute);
+
+        uint32_t x = compiler->get_execution_mode_argument(spv::ExecutionModeLocalSize, 0);
```
Comment (sjfricke) [inline layers/best_practices_utils.cpp:878]: I am very confused about why we need SPIR-V Cross to gather information from the SPIR-V

There is a `FindLocalSize` in the shader_validation file that gets the `ExecutionModeLocalSize`

There is nothing that should need SPIR-V Cross that can't be done from manually parsing out the SPIR-V Binary as the rest of the validation does.  Adding a 3rd party library instead of just using the infrastructure already for parsing this repo has seems like the wrong direction IMO

File: layers/best_practices_utils.cpp [hunk 2]
```diff
@@ -855,6 +859,105 @@ bool BestPractices::PreCallValidateCreateComputePipelines(VkDevice device, VkPip
...
+    }
+
+    return skip;
+}
+
+bool BestPractices::ValidateCreateComputePipelineArm(const VkComputePipelineCreateInfo& createInfo) const {
+    bool skip = false;
+    auto* module = GetShaderModuleState(createInfo.stage.module);
+    try {
+        auto compiler = std::unique_ptr<spirv_cross::Compiler>(new spirv_cross::Compiler(module->words));
```
Comment (sfricke-samsung) [inline layers/best_practices_utils.cpp:875]: I am very confused about why we need SPIR-V Cross to gather information from the SPIR-V

There is a `FindLocalSize` in the shader_validation file that gets the `ExecutionModeLocalSize`

There is nothing that should need SPIR-V Cross that can't be done from manually parsing out the SPIR-V Binary as the rest of the validation does.  Adding a 3rd party library instead of just using the infrastructure already for parsing this repo has seems like the wrong direction IMO

Comment (samwalls-arm) [reply layers/best_practices_utils.cpp:875]: Hey there, @sfricke-samsung. Thanks for taking a look at this. Yes, I wasn't too sure about the existence of functionality in terms of SPIRV reflection when I started working on this, so I'm not adamant about adding dependencies if we don't need to.

I can see an easy existing way to get work-group local sizes, as you pointed out. Do you know of anything we have in the project which replicates the following SPIRV-Cross functionality, or could be modified to?
```cpp
// Generate warnings about work group sizes based on active resources.
auto active_variables = compiler->get_active_interface_variables();
auto resources = compiler->get_shader_resources(active_variables);
```
Tl;dr, one of the checks implemented in this PR relies on some notion of shader resources. SPIRV-Cross can give me a list of shader resources being used in a shader.

I can see something which _looks_ similar, `MarkAccessibleIds` (however, I'm just speculating). It's used in the following context in `state_tracker.cpp`
```cpp
void ValidationStateTracker::RecordPipelineShaderStage(VkPipelineShaderStageCreateInfo const *pStage, PIPELINE_STATE *pipeline,
                                                       PIPELINE_STATE::StageState *stage_state) const {
    // Validation shouldn't rely on anything in stage state being valid if the spirv isn't
    auto module = GetShaderModuleState(pStage->module);
    if (!module->has_valid_spirv) return;

    // Validation shouldn't rely on anything in stage state being valid if the entrypoint isn't present
    auto entrypoint = FindEntrypoint(module, pStage->pName, pStage->stage);
    if (entrypoint == module->end()) return;

    // Mark accessible ids
    stage_state->accessible_ids = MarkAccessibleIds(module, entrypoint);
    ProcessExecutionModes(module, entrypoint, pipeline);

    stage_state->descriptor_uses =
        CollectInterfaceByDescriptorSlot(module, stage_state->accessible_ids, &stage_state->has_writable_descriptor);
    // Capture descriptor uses for the pipeline
    for (auto use : stage_state->descriptor_uses) {
        // While validating shaders capture which slots are used by the pipeline
        const uint32_t slot = use.first.first;
        auto &reqs = pipeline->active_slots[slot][use.first.second];
        reqs = descriptor_req(reqs | DescriptorTypeToReqs(module, use.second.type_id));
        pipeline->max_active_slot = std::max(pipeline->max_active_slot, slot);
    }
}
```

Comment (samwalls-arm) [reply layers/best_practices_utils.cpp:875]: I've modified the checks so far to use the pre-existing `FindLocalSize`. Trying to figure out a way to replicate shader resource analysis.

Comment (samwalls-arm) [reply layers/best_practices_utils.cpp]: So, looking into SPIRV in more depth, and the existing SPIRV systems over the last few days, I think I've come up with something that solves my problem (more in the spirit of what's already here).

The key thing which I was looking for was a way to check if any of the actively used image resources in the shader are 2D. `MarkAccessibleIds` and `CollectInterfaceByDescriptorSlot` seem to succeed in this, in combination with a custom analysis function, `IsShaderResource2DArm`.

```cpp
    // Generate warnings about work group sizes based on active resources.
    auto entrypoint = FindEntrypoint(module, createInfo.stage.pName, createInfo.stage.stage);
    if (entrypoint == module->end()) return false;

    bool has_writeable_descriptors = false;
    auto accessible_ids = MarkAccessibleIds(module, entrypoint);
    auto descriptor_uses = CollectInterfaceByDescriptorSlot(module, accessible_ids, &has_writeable_descriptors);

    unsigned dimensions = 0;
    if (x > 1) dimensions++;
    if (y > 1) dimensions++;
    if (z > 1) dimensions++;
    // Here the dimension will really depend on the dispatch grid, but assume it's 1D.
    dimensions = std::max(dimensions, 1u);

    // If we're accessing images, we almost certainly want to have a 2D workgroup for cache reasons.
    // There are some false positives here. We could simply have a shader that does this within a 1D grid,
    // or we may have a linearly tiled image, but these cases are quite unlikely in practice.
    bool accesses_2d = false;
    for (const auto& usage : descriptor_uses) accesses_2d |= IsShaderResource2DArm(module, usage.second);
```

```cpp
static bool IsShaderResource2DArm(const SHADER_MODULE_STATE* module, const interface_var& resource) {
    auto type = module->get_def(resource.type_id);
    while (true) {
        switch (type.opcode()) {
            case spv::OpTypeArray:
            case spv::OpTypeRuntimeArray:
            case spv::OpTypeSampledImage:
                type = module->get_def(type.word(2));
                break;
            case spv::OpTypePointer:
                type = module->get_def(type.word(3));
                break;
            case spv::OpTypeImage: {
                auto dim = type.word(3);
                switch (dim) {
                    case spv::Dim1D:
                    case spv::DimBuffer:
                        return false;
                    default:
                        return true;
                }
            }
            default:
                return false;
        }
    }
}

```

@jeremy-lunarg @sfricke-samsung, what do you think? I've removed the dependency on `spirv_cross.hpp` in `best_practices_utils.cpp`. If this is all good, then we can remove the commit which adds SPIRV-Cross as a dependency.

**Comment by samwalls-arm** [comment]
That's spirv-cross expunged from these commits now. Awaiting more reviews when possible.

File: layers/best_practices_error_enums.h [hunk 3]
```diff
@@ -97,6 +97,12 @@ static const char DECORATE_UNUSED *kVUID_BestPractices_SuboptimalSwapchainImageC
 // Arm-specific best practice
 static const char DECORATE_UNUSED *kVUID_BestPractices_AllocateDescriptorSets_SuboptimalReuse =
     "UNASSIGNED-BestPractices-vkAllocateDescriptorSets-suboptimal-reuse";
+static const char DECORATE_UNUSED *kVUID_BestPractices_CreateComputePipelines_BadComputeThreadGroupAlignment =
```
Comment (sfricke-samsung) [inline layers/best_practices_error_enums.h:100]: nit: I would avoid start these with `bad` as no where else does and I could argue all best practice messages could be "bad" since its applied being int he best practice layer

File: layers/best_practices_utils.cpp [hunk 4]
```diff
@@ -866,6 +868,102 @@ bool BestPractices::PreCallValidateCreateComputePipelines(VkDevice device, VkPip
...
+    if (VendorCheckEnabled(kBPVendorArm)) {
+        for (size_t i = 0; i < createInfoCount; i++) {
+            skip |= ValidateCreateComputePipelineArm(pCreateInfos[i]);
+        }
+    }
+
+    return skip;
+}
+
+static bool IsShaderResource2DArm(const SHADER_MODULE_STATE* module, const interface_var& resource) {
```
Comment (sfricke-samsung) [inline layers/best_practices_utils.cpp:880]: Could we move this function to shader_validation.cpp and call it `IsShaderResource2D`

It is nice to keep all the shader stuff in single place, mainly if a new SPIR-V op comes out to add support for, I doubt people will come looking in the best_practices_util to find it... also this might be useful for other normal validation cases

File: layers/best_practices_utils.cpp [hunk 5]
```diff
@@ -866,6 +868,102 @@ bool BestPractices::PreCallValidateCreateComputePipelines(VkDevice device, VkPip
...
+    auto type = module->get_def(resource.type_id);
+    while (true) {
```
Comment (sfricke-samsung) [inline layers/best_practices_utils.cpp:882]: I don't think this `while(true)` loop is needed as it only loops once regardless

Comment (samwalls-arm) [reply layers/best_practices_utils.cpp:882]: Hey @sfricke-samsung, is it possible to have arbitrarily long indirection with a pointer type to a resource like this? If so, it might be best this way, otherwise I agree.

Comment (sfricke-samsung) [reply layers/best_practices_utils.cpp:882]: sorry, I totally misread the code... this loop looks fine to me

File: layers/best_practices_utils.cpp [hunk 6]
```diff
@@ -866,6 +868,102 @@ bool BestPractices::PreCallValidateCreateComputePipelines(VkDevice device, VkPip
...
+            case spv::OpTypeSampledImage:
+                type = module->get_def(type.word(2));
+                break;
+            case spv::OpTypePointer:
+                type = module->get_def(type.word(3));
+                break;
+            case spv::OpTypeImage: {
+                auto dim = type.word(3);
+                switch (dim) {
+                    case spv::Dim1D:
```
Comment (sfricke-samsung) [inline layers/best_practices_utils.cpp:893]: if the dim is `spv::DimCube` or `spv::Dim3D` is it suppose to return true?

Comment (samwalls-arm) [reply layers/best_practices_utils.cpp:893]: I think the function should in fact be `IsShaderResource1D`, and the return values swapped here, as 1D-ness is really all we need to know.

To also answer this question, which is related:
> Could we move this function to shader_validation.cpp and call it IsShaderResource2D

I took the logic regarding what counts as 1D directly from the current PerfDoc implementation. I think there may be a reason for that, relevant to the best advice for Arm. So, I'll ask around internally if there's any reason we shouldn't be more generic about this.

Comment (sfricke-samsung) [reply layers/best_practices_utils.cpp:893]: I think a better approuch might be calling it `uint32_t GetShaderImageDim()` and then having the ARM Best Practice logic just check if it is `spv::Dim1D` or whatever you want returned... this function is just doing basic spir-v parsing logic and is fully agnostic

File: layers/best_practices_error_enums.h (hunk 3 above)
Comment (samwalls-arm) [reply layers/best_practices_error_enums.h:100]: Agree with this. Modified it accordingly.

File: layers/best_practices_utils.cpp (hunk 6 above)
Comment (samwalls-arm) [reply layers/best_practices_utils.cpp:918]: I agree with that very strongly. However it's possible that the function I wrote should actually correspond to "counts as not-1D for the purposes of Arm's guidance on bad alignment to work groups", as opposed to "not-1D according to the SPIRV spec". I'll have to check whether those classes of shader resources are the same or not.

I can write some more generic helper code, including what you suggested, in any case. It may be useful for future contributors. Thanks!

Comment (samwalls-arm) [reply layers/best_practices_utils.cpp:918]: Checking around internally, `spv::DimBuffer` should indeed count as 1D for the purposes of this check, not just `spv::Dim1D`. This may differ from other vendors, or what might intuitively make sense from the SPIRV spec.

The comments about generic helper functions still stand. I'll work on them.

**Comment by jeremy-lunarg** [comment]
@samwalls-arm
Can you ping me when you think this is ready (or mark as WIP in the mean time)? Then, I'll run it through our CI one more time.

File: layers/best_practices_utils.cpp (hunk 6 above)
Comment (samwalls-arm) [reply layers/best_practices_utils.cpp]: I've added a little helper function now. `GetShaderResourceDimensionality`, returns `int32_t` corresponding to the `spv::Dim` value, if positive, or unknown/error if negative.

**Comment by jeremy-lunarg** [comment]
@samwalls-arm @sfricke-samsung 
Are you guys happy with this? I'll run it through our CI testing again.

**Comment by sfricke-samsung** [comment]
I have no more outstanding issues

**Comment by jeremy-lunarg** [comment]
@samwalls-arm 
The galaxy s10, nexus player, pixel 3a, and windows (amd and nvidia) are failing on `ComputeShaderBadWorkGroupThreadAlignmentTest`.

```
7-14 10:55:52.825 28056 28109 I VulkanLayerValidationTests: [ RUN      ] VkArmBestPracticesLayerTest.ComputeShaderBadWorkGroupThreadAlignmentTest
07-14 10:55:52.849 28056 28109 I vulkan  : Loaded layer VK_LAYER_KHRONOS_validation
07-14 10:55:52.999 28056 28109 W ValidationTest: 0xebadde09 skipped times: 0
--------- beginning of crash
07-14 10:55:53.000 28056 28109 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x10 in tid 28109 (Thread-6), pid 28056 (ValidationTests)
07-14 10:55:53.026  1971  1971 E crash_dump64: unknown process state: t
07-14 10:55:53.052  1971  1971 I crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstone
07-14 10:55:53.053  5560  5560 I /system/bin/tombstoned: received crash request for pid 28109
07-14 10:55:53.055  1971  1971 I crash_dump64: performing dump of process 28056 (target tid = 28109)
07-14 10:55:53.063  1971  1971 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
07-14 10:55:53.063  1971  1971 F DEBUG   : Build fingerprint: 'samsung/beyond1ltexx/beyond1:10/QP1A.190711.020/G973FXXS4BTB3:user/release-keys'
07-14 10:55:53.063  1971  1971 F DEBUG   : Revision: '26'
07-14 10:55:53.063  1971  1971 F DEBUG   : ABI: 'arm64'
07-14 10:55:53.064  1971  1971 F DEBUG   : Timestamp: 2020-07-14 10:55:53-0600
07-14 10:55:53.064  1971  1971 F DEBUG   : pid: 28056, tid: 28109, name: Thread-6  >>> com.example.VulkanLayerValidationTests <<<
07-14 10:55:53.064  1971  1971 F DEBUG   : uid: 11671
07-14 10:55:53.064  1971  1971 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x10
07-14 10:55:53.064  1971  1971 F DEBUG   : Cause: null pointer dereference
07-14 10:55:53.064  1971  1971 F DEBUG   :     x0  000000749feca020  x1  0000007484f71b70  x2  0000000000000000  x3  0000000000000000
07-14 10:55:53.064  1971  1971 F DEBUG   :     x4  0000000000000040  x5  0000000000000000  x6  0000000000000001  x7  0000000000000020
07-14 10:55:53.064  1971  1971 F DEBUG   :     x8  0000000000000010  x9  0000000000000001  x10 0000000000000506  x11 000000749feca148
07-14 10:55:53.064  1971  1971 F DEBUG   :     x12 000000749feca168  x13 00000074747d2000  x14 fffffffffc000000  x15 0000000000000050
07-14 10:55:53.064  1971  1971 F DEBUG   :     x16 00000074ed0f3c18  x17 000000759122a560  x18 000000749f890000  x19 0000007474e2ea40
07-14 10:55:53.064  1971  1971 F DEBUG   :     x20 0000007484f71b30  x21 0000007484f71c70  x22 00000074f5de0438  x23 0000007484f71b70
07-14 10:55:53.064  1971  1971 F DEBUG   :     x24 000000749fecb020  x25 0000000000000000  x26 000000007fffffff  x27 0000000000000001
07-14 10:55:53.064  1971  1971 F DEBUG   :     x28 00000074ecfd42b0  x29 00000074ecf7fb18
07-14 10:55:53.064  1971  1971 F DEBUG   :     sp  000000749fec9eb0  lr  00000074eb4088d4  pc  00000074eb4088e8
07-14 10:55:53.067  1971  1971 F DEBUG   : 
07-14 10:55:53.067  1971  1971 F DEBUG   : backtrace:
07-14 10:55:53.067  1971  1971 F DEBUG   :       #00 pc 0000000000aba8e8  /vendor/lib64/egl/libGLES_mali.so (BuildId: b82e320b3685019f4c9d9c2dcf6f6d5c)
07-14 10:55:53.067  1971  1971 F DEBUG   :       #01 pc 0000000000aba70c  /vendor/lib64/egl/libGLES_mali.so (vkCreateComputePipelines+556) (BuildId: b82e320b3685019f4c9d9c2dcf6f6d5c)
07-14 10:55:53.067  1971  1971 F DEBUG   :       #02 pc 00000000008a6550  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVkLayer_khronos_validation.so (BuildId: 94793956deaa937900fc4b9e4361da1a4ccff34d)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #03 pc 00000000008e8478  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVkLayer_khronos_validation.so (BuildId: 94793956deaa937900fc4b9e4361da1a4ccff34d)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #04 pc 000000000001f228  /system/lib64/libvulkan.so (vulkan::api::(anonymous namespace)::CreateComputePipelines(VkDevice_T*, VkPipelineCache_T*, unsigned int, VkComputePipelineCreateInfo const*, VkAllocationCallbacks const*, VkPipeline_T**)+224) (BuildId: d569637329ba16330d390540dee0e601)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #05 pc 00000000003290f8  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #06 pc 00000000003d8904  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (testing::Test::Run()+664) (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #07 pc 00000000003d9f34  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (testing::TestInfo::Run()+672) (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #08 pc 00000000003da4fc  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (testing::TestSuite::Run()+272) (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #09 pc 00000000003e8580  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (testing::internal::UnitTestImpl::RunAllTests()+1204) (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #10 pc 00000000003e7f90  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (testing::UnitTest::Run()+124) (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #11 pc 000000000020a2c8  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #12 pc 00000000003c5b18  /data/app/com.example.VulkanLayerValidationTests-LWduv5EKesrfGaS_f-_9QQ==/lib/arm64/libVulkanLayerValidationTests.so (BuildId: 8e46a192d941ff7e362ecfbda11e5df32ba863e9)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #13 pc 00000000000e3b14  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+36) (BuildId: 220051b49364b1c2da3adf10c30832cc)
07-14 10:55:53.068  1971  1971 F DEBUG   :       #14 pc 0000000000085330  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: 220051b49364b1c2da3adf10c30832cc)
```

**Comment by samwalls-arm** [comment]
@jeremy-lunarg, that's quite odd...

Just to confirm, do the other two compute tests pass?

**Comment by jeremy-lunarg** [comment]
> @jeremy-lunarg, that's quite odd...
> 
> Just to confirm, do the other two compute tests pass?

Here's the entire log for the galaxy device.

[galaxy.txt](https://github.com/KhronosGroup/Vulkan-ValidationLayers/files/4926606/galaxy.txt)

**Comment by samwalls-arm** [comment]
All three tests pass on both my local machines (nvidia, linux) + (intel graphics, windows), as well as a colleague's machine (nvidia, windows). I've yet to try this out on an Android build. However, given this crashes for your CI's windows setup as well, on AMD and Nvidia, I'm really not sure what's going on.

How is the internal CI built/run?

**Comment by samwalls-arm** [comment]
Hey, @jeremy-lunarg, do you have an update on this?

**Comment by jeremy-lunarg** [comment]
> Hey, @jeremy-lunarg, do you have an update on this?

Yessir, I reran the tests. Still seeing those failures. I'm looking into right now.

**Comment by jeremy-lunarg** [comment]
I've asked Mark to give us a hand with this one. He's good at hunting done the android issues and I'm going to be out of town for a few days.

**Comment by mark-lunarg** [comment]
Hey @samwalls-arm, after rebasing, this is still crashing on the AMD Windows platforms and all of our Android hardware. We don't have the time to debug PR issues (we hardly do for ourselves!), but as it also fails on _my_ android hardware, I'll spend some time trying to narrow down the failure and get back to you.

**Comment by samwalls-arm** [comment]
Thanks for looking into this! I've got everything building on android now, and can finally confirm the same behaviour on my end.

My team will be moving onto other projects as our primary work, very soon. So, this may take a while for us to resolve. However, if you do manage to narrow anything down, let me know what you find, and I may be able to take a look in the future, since it's probably not a major problem (I can only hope).

**Comment by Tony-LunarG** [comment]
I fixed the tests, shortened one of the commit messages, rebased and merged to master

**Comment by samwalls-arm** [comment]
@Tony-LunarG ah that's great! Thank you very much for looking into this.

For my own sanity, what was wrong with the existing tests?

**Comment by Tony-LunarG** [comment]
ComputePipelineCreateInfo.layout was 0.  I added "pipe.cp_ci_.layout = pipe.pipeline_layout_.handle();" to the two make_pipeline_with_shader functions.

**Comment by mark-lunarg** [comment]
@samwalls-arm -- @Tony-LunarG said this error was also caught by running validation (full) on the best_practices tests!
